cli/azd/pkg/async/task_test.go

```diff
@@ -0,0 +1,330 @@
+				DefaultValue: "eastus2",
+			})
+
+			if err != nil {
+				return err
+			}
+
+			confirm, err := console.Confirm(ctx, input.ConsoleOptions{
+				Message:      "Are you sure?",
+				DefaultValue: true,
+			})
+
+			if err != nil {
+				return err
+			}
+
+			if !confirm {
+				return errors.New("User did not confirm")
+			}
+
+			selectedLocation = location
+
+			return nil
+		})
+
+		if err != nil {
+			taskContext.SetError(err)
+			return
+		}
+
+		taskContext.SetResult(selectedLocation)
+	})
+
+	err := task.Run()
+	require.NoError(t, err)
+
+	go func() {
+		for status := range task.Progress() {
+			progress = append(progress, status)
+		}
+	}()
+
+	go func() {
+		for isInteractive := range task.interactiveChannel {
+			interactiveStatus = append(interactiveStatus, isInteractive)
+		}
+	}()
+
+	actualResult, err := task.Await()
+	// Result still expected
+	require.Equal(t, expectedResult, actualResult)
+	require.Nil(t, err)
+	// Progress still reported
+	require.Equal(t, 2, len(progress))
+	require.Equal(t, "thing 1", progress[0])
+	require.Equal(t, "thing 2", progress[1])
+	// interactive status reported
+	require.Equal(t, 2, len(interactiveStatus))
+	require.Equal(t, true, interactiveStatus[0])
+	require.Equal(t, false, interactiveStatus[1])
+}
+
+func TestInteractiveTaskWithError(t *testing.T) {
+	ctx := context.Background()
+	console := mocks.NewMockConsole()
+	progress := []string{}
+	interactiveStatus := []bool{}
+	expectedError := errors.New("User did not confirm")
+
+	console.WhenPrompt(func(options input.ConsoleOptions) bool {
+		return options.Message == "What location?"
+	}).Respond("westus2")
+
+	// This time the user will not confirm
+	console.WhenConfirm(func(options input.ConsoleOptions) bool {
+		return options.Message == "Are you sure?"
+	}).Respond(false)
+
+	task := NewInteractiveTaskWithProgress(func(taskContext *InteractiveTaskContextWithProgress[string, string]) {
+		var selectedLocation string
+
+		taskContext.SetProgress("thing 1")
+		taskContext.SetProgress("thing 2")
+
+		err := taskContext.Interact(func() error {
+			_, err := console.Prompt(ctx, input.ConsoleOptions{
+				Message:      "What location?",
+				DefaultValue: "eastus2",
+			})
+
+			if err != nil {
+				return err
+			}
+
+			confirm, err := console.Confirm(ctx, input.ConsoleOptions{
+				Message:      "Are you sure?",
+				DefaultValue: true,
+			})
+
+			if err != nil {
+				return err
+			}
+
+			if !confirm {
+				return expectedError
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			taskContext.SetError(err)
+			return
+		}
+
+		taskContext.SetResult(selectedLocation)
+	})
+
+	err := task.Run()
+	require.NoError(t, err)
+
+	go func() {
+		for status := range task.Progress() {
+			progress = append(progress, status)
+		}
+	}()
+
+	go func() {
+		for isInteractive := range task.interactiveChannel {
+			interactiveStatus = append(interactiveStatus, isInteractive)
+		}
+	}()
+
+	actualResult, err := task.Await()
```

Comment:
Looking at the test failure, we might need to additionally do two things:
1. The task completion (`Await()`, `Result()`) should be closing all channels (interactive,Channel progressChannel). Otherwise, listeners will wait forever for the next channel item.
2. We need to wait for the listener goroutines to shut down before proceeding with assertion. (Similar to #223)